### ISSUE TITLE
Incus: Creating new containers/vms blocked by apparmor

### ISSUE THREAD

**Issue by broizter** [NixOS/nixpkgs#350012]
When creating new containers or VMs using Incus you first have to disable apparmor, otherwise you get errors and the creation fails.

```bash
incus create images:debian/12 test
Creating test
Error: Failed instance creation: Failed creating instance from image: Unpack failed: Failed to run: tar --anchored --wildcards --exclude=dev/* --exclude=/dev/* --exclude=./dev/* --exclude=rootfs/dev/* --exclude=/rootfs/dev/* --exclude=./rootfs/dev/* --restrict --force-local -C /var/lib/incus/storage-pools/default/images/149e0ff3a1ca46db20336210704104db4c9409f2f33369dea26174d908757c29 --numeric-owner --xattrs-include=* -Jxf -: exit status 127 (tar: error while loading shared libraries: libacl.so.1: cannot open shared object file: No such file or directory)

dmesg -T
[sön okt 20 14:32:58 2024] audit: type=1400 audit(1729427573.016:126): apparmor="STATUS" operation="profile_load" profile="unconfined" name="incus_archive-d6df5eea-1a10-4ad0-b163-43dbbc49e4b1" pid=2582270 comm="apparmor_parser"
[sön okt 20 14:32:58 2024] audit: type=1400 audit(1729427573.018:127): apparmor="DENIED" operation="open" class="file" profile="incus_archive-d6df5eea-1a10-4ad0-b163-43dbbc49e4b1" name="/nix/store/7px4n99mcmdzx8nygx59f28j8g7vj0kb-acl-2.3.2/lib/libacl.so.1.1.2302" pid=2582272 comm="tar" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
[sön okt 20 14:32:58 2024] audit: type=1400 audit(1729427573.057:128): apparmor="STATUS" operation="profile_remove" profile="unconfined" name="incus_archive-d6df5eea-1a10-4ad0-b163-43dbbc49e4b1" pid=2582276 comm="apparmor_parser"
```

**Comment by broizter** [comment]
The maintainer of Incus [Stéphane Graber](https://discuss.linuxcontainers.org/u/stgraber) wrote about the issue which I thought was worth attaching.

"So normally the AppArmor abstractions/base contains what’s needed for dynamic libraries to be loaded. All our profiles include that abstraction so we don’t need to specifically allow every single library.

There must be something odd going on with NixOS which causes that abstraction to not include a correct pattern for where Nix stores its libraries (/nix/storage/*/lib/so or something like that)."

**Comment by broizter** [comment]
This is the apparmor profile he suggested to fix this particular issue.

```
/nix/storage/*/lib/*so* mr,
```

**Comment by umbra** [comment]
Disabling apparmor, creating the container, then re-enabling apparmor is a (brute force, totally stupid) work around.

The above should be (sic by Stéphane Graber):

`/nix/store/*/lib/*so* mr,`

However, I could not figure out how to install that as a profile on NixOS.

**Comment by broizter** [comment]
> However, I could not figure out how to install that as a profile on NixOS.

Yeah I couldn't figure this out either. I had to move away from NixOS because of this problem. It was a bit too annoying for a server which main purpose was running Incus.

**Comment by MakiseKurisu** [comment]
@NixOS/lxc Maybe someone can take a look at this issue? I tried [some AppArmor profiles](https://discuss.linuxcontainers.org/t/creating-new-containers-vms-blocked-by-apparmor-on-nixos/21908/8?u=makisekurisu) but they do not apply to the gnutar launched by incusd.

**Comment by adamcstephens** [comment]
AppArmor is off by default on NixOS. When I investigated using it a long time ago, it seemed far from functional. Somebody will need to put in the effort to work through the problems, but that clearly hasn’t happened.

I don’t know AppArmor well enough at this point to even know what’s necessary. I’d be happy to review changes necessary to make this work if you can figure out the issues. Any fixes should end up in a nixos test to ensure it keeps working moving forward.

**Comment by MakiseKurisu** [comment]
Ping @ju1m and @LordGrimmauld who are the AA maintainers in nixpkgs.
